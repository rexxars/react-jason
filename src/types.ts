export type TokenType =
  | 'array'
  | 'attribute'
  | 'boolean'
  | 'nil'
  | 'number'
  | 'object'
  | 'punctuation'
  | 'quotation'
  | 'root'
  | 'string'
  | 'unquotedAttribute'

export type TokenMachine = (token: TokenType, content?: React.ReactNode, as?: string) => JSX.Element
export type ItemKeyGenerator = (item: unknown, basePath: string, index: number) => string
export type NodeWrapper = React.ComponentType<{path: string}>

export type JasonStyles = Partial<Record<TokenType, React.CSSProperties>>
export type JasonClasses = Partial<Record<TokenType, string>>

export interface JasonTheme {
  styles?: JasonStyles
  classes?: JasonClasses
}

export interface JasonContextInstance {
  token: TokenMachine
  getItemKey: ItemKeyGenerator
  nodeWrapper?: NodeWrapper
  quoteAttributes: boolean
}
