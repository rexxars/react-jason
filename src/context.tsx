import React, {createContext, createElement, useCallback, useContext} from 'react'
import {JasonContextInstance, JasonTheme, TokenMachine, TokenType} from './types'

// We're always rendering the context with values, thus "any"
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const JasonContext = createContext<JasonContextInstance>({} as any)

export function useTokenMachine(): {token: TokenMachine; char: (sym: string) => JSX.Element} {
  const {token} = useContext(JasonContext)
  const char = useCallback(
    (sym: string) => token(sym === '"' || sym === "'" ? 'quotation' : 'punctuation', sym),
    [token],
  )
  return {token, char}
}

export function useToken(token: TokenType, children?: React.ReactNode, as = 'span'): JSX.Element {
  return useTokenMachine().token(token, children, as)
}

export function createTokenMachine(theme: JasonTheme): TokenMachine {
  return function tokenMachine(
    token: TokenType,
    children?: React.ReactNode,
    as = 'span',
  ): JSX.Element {
    const style = theme.styles ? theme.styles[token] : undefined
    const className = theme.classes ? theme.classes[token] : undefined
    const styled = Boolean(style || className)
    return styled ? createElement(as, {style, className}, children) : <>{children}</>
  }
}
