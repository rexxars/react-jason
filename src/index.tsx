export * from './react-jason'
export * from './types'
export * from './paths'
