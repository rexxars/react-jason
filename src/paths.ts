/* eslint-disable @typescript-eslint/ban-ts-comment */
import {ItemKeyGenerator} from './types'

export const defaultItemKeyGenerator: ItemKeyGenerator = (
  // @ts-ignore
  item: unknown,
  basePath: string,
  index: number,
) => {
  return `${basePath}[${index}]`
}

export const sanityItemKeyGenerator: ItemKeyGenerator = (
  item: unknown,
  basePath: string,
  index: number,
) => {
  return isKeyedObject(item) ? `${basePath}[_key=="${item._key}"]` : `${basePath}[${index}]`
}

function isObject(obj: unknown): obj is Record<string, unknown> {
  return typeof obj === 'object' && obj !== null && !Array.isArray(obj)
}

function isKeyedObject(obj: unknown): obj is {_key: string} {
  return isObject(obj) && typeof obj._key === 'string'
}
