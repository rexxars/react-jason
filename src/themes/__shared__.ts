import {CSSProperties} from 'react'

export const sharedRoot: CSSProperties = {
  fontFamily: 'Menlo, Monaco, "Courier New", monospace',
  fontFeatureSettings: '"liga" 0, "calt" 0',
  lineHeight: '1.5em',
  whiteSpace: 'pre',
  margin: 0,
}
