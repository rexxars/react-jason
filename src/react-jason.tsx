import React, {useContext, Fragment} from 'react'
import {JasonContext} from './context'
import {defaultItemKeyGenerator} from './paths'
import {defaultTheme} from './themes'
import {createTokenMachine, useToken, useTokenMachine} from './context'
import {JasonTheme, JasonContextInstance, ItemKeyGenerator, NodeWrapper as INodeWrapper} from './types'

export * as themes from './themes'

export interface JasonProps {
  value: unknown
  theme?: JasonTheme
  nodeWrapper?: INodeWrapper
  quoteAttributes?: boolean
  itemKeyGenerator?: ItemKeyGenerator
}

export const ReactJason = ({
  value,
  theme,
  nodeWrapper,
  itemKeyGenerator,
  quoteAttributes = true,
}: JasonProps) => {
  const token = createTokenMachine(theme || defaultTheme)
  const getItemKey = itemKeyGenerator || defaultItemKeyGenerator
  const context = React.useMemo<JasonContextInstance>(
    () => ({token, getItemKey, quoteAttributes, nodeWrapper}),
    [token, getItemKey, quoteAttributes, nodeWrapper],
  )

  return (
    <JasonContext.Provider value={context}>
      {token('root', node({value, path: '', depth: 1, nodeWrapper}), 'pre')}
    </JasonContext.Provider>
  )
}

function node({
  value,
  path,
  depth,
  nodeWrapper: NodeWrapper,
}: {
  value: unknown
  path: string
  depth: number
  nodeWrapper?: INodeWrapper
}): JSX.Element {
  const wrap = (children: React.ReactElement) =>
    NodeWrapper ? <NodeWrapper path={path}>{children}</NodeWrapper> : children

  if (value === null) {
    return wrap(<NullNode />)
  }

  if (Array.isArray(value)) {
    return wrap(<ArrayNode value={value} path={path} depth={depth} />)
  }

  switch (typeof value) {
    case 'string':
      return wrap(<StringNode value={value} path={path} depth={depth} />)
    case 'number':
      return wrap(<NumberNode value={value} path={path} depth={depth} />)
    case 'boolean':
      return wrap(<BooleanNode value={value} path={path} depth={depth} />)
    case 'object':
      return wrap(<ObjectNode value={value as Record<string, unknown>} path={path} depth={depth} />)
    default:
      throw new Error(`Unhandled type ${typeof value}`)
  }
}

function StringNode({value}: {value: string; path: string; depth: number}) {
  const {char} = useTokenMachine()
  const unquoted = json(value).slice(1, -1)
  return useToken(
    'string',
    <>
      {char('"')}
      {unquoted}
      {char('"')}
    </>,
  )
}

function NumberNode({value}: {value: number; path: string; depth: number}) {
  return useToken('number', value)
}

function BooleanNode({value}: {value: boolean; path: string; depth: number}) {
  return useToken('boolean', value ? 'true' : 'false')
}

function ArrayNode({value, path, depth}: {value: unknown[]; path: string; depth: number}) {
  const {token, char} = useTokenMachine()
  const {getItemKey, nodeWrapper} = useContext(JasonContext)
  const numItems = value.length
  return token(
    'array',
    <>
      {char('[')}
      {`\n${indent(depth)}`}
      {value.map((item, index) => {
        const itemPath = getItemKey(item, path, index)
        const isLast = index === numItems - 1
        return (
          <Fragment key={itemPath}>
            {node({value: item, path: itemPath, depth: depth + 1, nodeWrapper})}
            {isLast ? (
              `\n${indent(depth - 1)}`
            ) : (
              <>
                {char(',')}
                {`\n${indent(depth)}`}
              </>
            )}
          </Fragment>
        )
      })}
      {char(']')}
    </>,
  )
}

function ObjectNode({value: obj, path, depth}: {value: Record<string, unknown>; path: string; depth: number}) {
  const {nodeWrapper} = useContext(JasonContext)
  const {token, char} = useTokenMachine()
  const value = obj as Record<string, unknown>
  const keys = Object.keys(value)
  const lastKey = keys.length - 1
  return token(
    'object',
    <>
      {char('{')}
      {'\n'}
      {keys.map((key, index) => {
        const val = value[key]
        const propPath = path ? `${path}.${key}` : key
        return (
          <Fragment key={propPath}>
            {indent(depth)}
            <AttributeNode value={key}/>
            {node({value: val, path: propPath, depth: depth + 1, nodeWrapper})}
            {index < lastKey ? (
              <>
                {char(',')}
                {'\n'}
              </>
            ) : (
              '\n'
            )}
          </Fragment>
        )
      })}
      {indent(depth - 1)}
      {char('}')}
    </>,
  )
}

function AttributeNode({value}: {value: string}) {
  const {quoteAttributes} = useContext(JasonContext)
  const {token, char} = useTokenMachine()
  const attributeName = json(value).slice(1, -1)

  let attribute: React.ReactNode
  if (quoteAttributes || attributeName.includes('"')) {
    attribute = token(
      'attribute',
      <>
        {char('"')}
        {attributeName}
        {char('"')}
      </>,
    )
  } else {
    attribute = token('unquotedAttribute', attributeName)
  }

  return (
    <>
      {attribute}
      {char(':')}{' '}
    </>
  )
}

function NullNode() {
  return useToken('nil', 'null')
}

function json(value: unknown) {
  return JSON.stringify(value)
}

function indent(depth: number): string {
  return '  '.repeat(depth)
}
