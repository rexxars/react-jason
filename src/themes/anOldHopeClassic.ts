import {JasonTheme} from '../types'
import {sharedRoot} from './__shared__'

const theme: JasonTheme = {
  styles: {
    root: Object.assign({}, sharedRoot, {backgroundColor: '#1e2024'}),
    attribute: {color: '#ef415a'},
    unquotedAttribute: {color: '#d1d3d7'},
    string: {color: '#57bedf'},
    nil: {color: '#f2851e'},
    number: {color: '#f2851e'},
    boolean: {color: '#f2851e'},
    punctuation: {color: '#d1d3d7'},
  },
}

export default theme
