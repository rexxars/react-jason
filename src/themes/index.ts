import anOldHopeClassic from './anOldHopeClassic'

export {default as anOldHopeClassic} from './anOldHopeClassic'
export {default as github} from './github'
export {default as monokai} from './monokai'
export {default as monokaiSublime} from './monokaiSublime'
export {default as vscodeDark} from './vscodeDark'
export {default as vscodeLight} from './vscodeLight'

export const defaultTheme = anOldHopeClassic
