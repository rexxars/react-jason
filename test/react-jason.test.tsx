/* eslint-disable react/jsx-no-bind */
import * as React from 'react'
import {render, unmountComponentAtNode} from 'react-dom'
import {renderToStaticMarkup} from 'react-dom/server'
import renderer from 'react-test-renderer'
import {github} from '../src/themes'
import {JasonTheme, NodeWrapper} from '../src/types'
import {ReactJason} from '../src/react-jason'
import {sanityItemKeyGenerator} from '../src/paths'

const value = {
  string: 'value',
  number: 13.37,
  bool: true,
  nil: null,
  array: ['primitive', 'values'],
  object: {nested: {arbitrarily: {deep: true}}},
  objectArray: [
    {_key: 'grrm', name: 'George. R. R. Martin'},
    {_key: 'jrrt', name: 'J. R. R. Tolkien'},
  ],
}

const classTheme: JasonTheme = {
  classes: {
    root: 'root',
    attribute: 'attribute',
    unquotedAttribute: 'unquotedAttribute',
    string: 'string',
    nil: 'nil',
    number: 'number',
    boolean: 'boolean',
    punctuation: 'punctuation',
  },
}

describe('react-jason', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div')
    render(<ReactJason value={value} />, div)
    unmountComponentAtNode(div)
  })

  it('renders expected values', () => {
    const component = renderer.create(<ReactJason value={value} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('can specify theme', () => {
    const defaultTheme = renderer.create(<ReactJason value={value} />).toJSON() as any
    const githubTheme = renderer.create(<ReactJason value={value} theme={github} />).toJSON()
    expect(githubTheme).toMatchSnapshot()
    expect(githubTheme).not.toMatchObject(defaultTheme)
  })

  it('can use class-based theme', () => {
    const classBased = renderer.create(<ReactJason value={value} theme={classTheme} />).toJSON()
    expect(classBased).toMatchSnapshot()
  })

  it('can render unquotes props', () => {
    const unquoted = renderToStaticMarkup(<ReactJason value={value} quoteAttributes={false} />)
    expect(unquoted).not.toMatch('&quot;string&quot;')

    const quoted = renderToStaticMarkup(<ReactJason value={value} quoteAttributes />)
    expect(quoted).toMatch('&quot;string&quot;')
  })

  it('can use custom node wrapper', () => {
    const nodeWrapper: NodeWrapper = ({path, children}) => <span data-path={path}>{children}</span>
    const wrapped = renderToStaticMarkup(<ReactJason value={value} nodeWrapper={nodeWrapper} />)
    expect(wrapped).toMatch('data-path="objectArray[1].name"')
  })

  it('can use custom item key generator', () => {
    const nodeWrapper: NodeWrapper = ({path, children}) => <span data-path={path}>{children}</span>
    const wrapped = renderToStaticMarkup(
      <ReactJason
        value={value}
        nodeWrapper={nodeWrapper}
        itemKeyGenerator={sanityItemKeyGenerator}
      />,
    )
    expect(wrapped).not.toMatch('data-path="objectArray[1].name"')
    expect(wrapped).toMatch('data-path="objectArray[_key==&quot;grrm&quot;].name"')
  })
})
