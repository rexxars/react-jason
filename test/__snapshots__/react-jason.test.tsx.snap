// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-jason can specify theme 1`] = `
<pre
  style={
    Object {
      "backgroundColor": "#f8f8f8",
      "fontFamily": "Menlo, Monaco, \\"Courier New\\", monospace",
      "fontFeatureSettings": "\\"liga\\" 0, \\"calt\\" 0",
      "lineHeight": "1.5em",
      "margin": 0,
      "whiteSpace": "pre",
    }
  }
>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    {
  </span>
  

    
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    string
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#de1044",
      }
    }
  >
    "
    value
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    ,
  </span>
  

    
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    number
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#008080",
      }
    }
  >
    13.37
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    ,
  </span>
  

    
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    bool
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#008080",
      }
    }
  >
    true
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    ,
  </span>
  

    
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    nil
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#008080",
      }
    }
  >
    null
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    ,
  </span>
  

    
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    array
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    [
  </span>
  
    
  <span
    style={
      Object {
        "color": "#de1044",
      }
    }
  >
    "
    primitive
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    ,
  </span>
  
    
  <span
    style={
      Object {
        "color": "#de1044",
      }
    }
  >
    "
    values
    "
  </span>
  
  
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    ]
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    ,
  </span>
  

    
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    object
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    {
  </span>
  

      
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    nested
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    {
  </span>
  

        
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    arbitrarily
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    {
  </span>
  

          
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    deep
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#008080",
      }
    }
  >
    true
  </span>
  

        
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    }
  </span>
  

      
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    }
  </span>
  

    
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    }
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    ,
  </span>
  

    
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    objectArray
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    [
  </span>
  
    
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    {
  </span>
  

        
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    _key
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#de1044",
      }
    }
  >
    "
    grrm
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    ,
  </span>
  

        
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    name
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#de1044",
      }
    }
  >
    "
    George. R. R. Martin
    "
  </span>
  

      
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    }
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    ,
  </span>
  
    
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    {
  </span>
  

        
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    _key
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#de1044",
      }
    }
  >
    "
    jrrt
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    ,
  </span>
  

        
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    "
    name
    "
  </span>
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#de1044",
      }
    }
  >
    "
    J. R. R. Tolkien
    "
  </span>
  

      
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    }
  </span>
  
  
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    ]
  </span>
  

  
  <span
    style={
      Object {
        "color": "#333",
      }
    }
  >
    }
  </span>
</pre>
`;

exports[`react-jason can use class-based theme 1`] = `
<pre
  className="root"
>
  <span
    className="punctuation"
  >
    {
  </span>
  

    
  <span
    className="attribute"
  >
    "
    string
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="string"
  >
    "
    value
    "
  </span>
  <span
    className="punctuation"
  >
    ,
  </span>
  

    
  <span
    className="attribute"
  >
    "
    number
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="number"
  >
    13.37
  </span>
  <span
    className="punctuation"
  >
    ,
  </span>
  

    
  <span
    className="attribute"
  >
    "
    bool
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="boolean"
  >
    true
  </span>
  <span
    className="punctuation"
  >
    ,
  </span>
  

    
  <span
    className="attribute"
  >
    "
    nil
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="nil"
  >
    null
  </span>
  <span
    className="punctuation"
  >
    ,
  </span>
  

    
  <span
    className="attribute"
  >
    "
    array
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="punctuation"
  >
    [
  </span>
  
    
  <span
    className="string"
  >
    "
    primitive
    "
  </span>
  <span
    className="punctuation"
  >
    ,
  </span>
  
    
  <span
    className="string"
  >
    "
    values
    "
  </span>
  
  
  <span
    className="punctuation"
  >
    ]
  </span>
  <span
    className="punctuation"
  >
    ,
  </span>
  

    
  <span
    className="attribute"
  >
    "
    object
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="punctuation"
  >
    {
  </span>
  

      
  <span
    className="attribute"
  >
    "
    nested
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="punctuation"
  >
    {
  </span>
  

        
  <span
    className="attribute"
  >
    "
    arbitrarily
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="punctuation"
  >
    {
  </span>
  

          
  <span
    className="attribute"
  >
    "
    deep
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="boolean"
  >
    true
  </span>
  

        
  <span
    className="punctuation"
  >
    }
  </span>
  

      
  <span
    className="punctuation"
  >
    }
  </span>
  

    
  <span
    className="punctuation"
  >
    }
  </span>
  <span
    className="punctuation"
  >
    ,
  </span>
  

    
  <span
    className="attribute"
  >
    "
    objectArray
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="punctuation"
  >
    [
  </span>
  
    
  <span
    className="punctuation"
  >
    {
  </span>
  

        
  <span
    className="attribute"
  >
    "
    _key
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="string"
  >
    "
    grrm
    "
  </span>
  <span
    className="punctuation"
  >
    ,
  </span>
  

        
  <span
    className="attribute"
  >
    "
    name
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="string"
  >
    "
    George. R. R. Martin
    "
  </span>
  

      
  <span
    className="punctuation"
  >
    }
  </span>
  <span
    className="punctuation"
  >
    ,
  </span>
  
    
  <span
    className="punctuation"
  >
    {
  </span>
  

        
  <span
    className="attribute"
  >
    "
    _key
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="string"
  >
    "
    jrrt
    "
  </span>
  <span
    className="punctuation"
  >
    ,
  </span>
  

        
  <span
    className="attribute"
  >
    "
    name
    "
  </span>
  <span
    className="punctuation"
  >
    :
  </span>
   
  <span
    className="string"
  >
    "
    J. R. R. Tolkien
    "
  </span>
  

      
  <span
    className="punctuation"
  >
    }
  </span>
  
  
  <span
    className="punctuation"
  >
    ]
  </span>
  

  
  <span
    className="punctuation"
  >
    }
  </span>
</pre>
`;

exports[`react-jason renders expected values 1`] = `
<pre
  style={
    Object {
      "backgroundColor": "#1e2024",
      "fontFamily": "Menlo, Monaco, \\"Courier New\\", monospace",
      "fontFeatureSettings": "\\"liga\\" 0, \\"calt\\" 0",
      "lineHeight": "1.5em",
      "margin": 0,
      "whiteSpace": "pre",
    }
  }
>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    {
  </span>
  

    
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    string
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#57bedf",
      }
    }
  >
    "
    value
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    ,
  </span>
  

    
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    number
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#f2851e",
      }
    }
  >
    13.37
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    ,
  </span>
  

    
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    bool
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#f2851e",
      }
    }
  >
    true
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    ,
  </span>
  

    
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    nil
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#f2851e",
      }
    }
  >
    null
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    ,
  </span>
  

    
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    array
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    [
  </span>
  
    
  <span
    style={
      Object {
        "color": "#57bedf",
      }
    }
  >
    "
    primitive
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    ,
  </span>
  
    
  <span
    style={
      Object {
        "color": "#57bedf",
      }
    }
  >
    "
    values
    "
  </span>
  
  
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    ]
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    ,
  </span>
  

    
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    object
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    {
  </span>
  

      
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    nested
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    {
  </span>
  

        
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    arbitrarily
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    {
  </span>
  

          
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    deep
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#f2851e",
      }
    }
  >
    true
  </span>
  

        
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    }
  </span>
  

      
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    }
  </span>
  

    
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    }
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    ,
  </span>
  

    
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    objectArray
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    [
  </span>
  
    
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    {
  </span>
  

        
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    _key
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#57bedf",
      }
    }
  >
    "
    grrm
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    ,
  </span>
  

        
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    name
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#57bedf",
      }
    }
  >
    "
    George. R. R. Martin
    "
  </span>
  

      
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    }
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    ,
  </span>
  
    
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    {
  </span>
  

        
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    _key
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#57bedf",
      }
    }
  >
    "
    jrrt
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    ,
  </span>
  

        
  <span
    style={
      Object {
        "color": "#ef415a",
      }
    }
  >
    "
    name
    "
  </span>
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    :
  </span>
   
  <span
    style={
      Object {
        "color": "#57bedf",
      }
    }
  >
    "
    J. R. R. Tolkien
    "
  </span>
  

      
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    }
  </span>
  
  
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    ]
  </span>
  

  
  <span
    style={
      Object {
        "color": "#d1d3d7",
      }
    }
  >
    }
  </span>
</pre>
`;
